### DISCUSSION TITLE
Performance issues while reading TimeBase data

### DISCUSSION THREAD

**Discussion by bardale** [finos/TimeBase-CE#214] [Q&A]
Hello,

We are requesting data from TimeBase server and we see below stats for fetching data.

**Time required to fetch 30 mins timespan data:**

Time taken to open connection: 8581 ms
Time taken to open stream: 557 ms
Time taken to initialize decoder: 10 ms
Total Time taken (opening connection and read data) to fetch records from timebase server: 13455 ms
Total number of records fetched: 17477

**Time required to fetch 60 mins timespan data:**

Time taken to open connection: 8988 ms
Time taken to open stream: 600 ms
Time taken to initialize decoder:  20 ms
Total Time taken (opening connection and read data) to fetch records from timebase server: 15792 ms
Total number of records fetched: 38318

**Please find below code used to load timebase:**

------ QQL for creating Stream-----
CREATE Durable STREAM "bbo_stream" (
	CLASS "test.manual.message.BBOMessageClass" (
		"tradetype" 'TradeType' Varchar,
		"bestbidprice" 'TradeType' FLOAT DECIMAL64,
		"bestaskprice" 'bestaskprice' FLOAT DECIMAL64,
		"bidsize" 'bidsize' FLOAT DECIMAL64,
		"asksize" 'asksize' FLOAT DECIMAL64,
		"exch" 'Exch' Varchar
	);
)
OPTIONS (FIXEDTYPE; PERIODICITY = 'IRREGULAR'; HIGHAVAILIBILITY = TRUE)
COMMENT 'BestBidOffer Data'

**python code to load Timebase stream**

db = tbapi.TickDb.createFromUrl(timebaseUrl)
db.open(False)
  
streamKey = 'bbo_stream'
stream = createStreamQql(db, streamKey) #this function calls file which contains QQL of "bbo_stream"

options = tbapi.LoadingOptions()
options.WriteMode = tbapi.WriteMode('APPEND')	

loader = stream.createLoader(options)

bboMessage = tbapi.InstrumentMessage()
bboMessage.typeName = 'test.manual.message.BBOMessageClass'

**using pandas to read from file and load data to TimeBase  
iterating pandas dataframe to insert each row into BBOMessageClass attributes and then loading to TimeBase using**
loader.send(bboMessage)

loader.close()
db.close()

**Code to read data from Timebase server:** 
long logStartTime = System.currentTimeMillis();
DXTickDB db = TickDBFactory.createFromUrl (timebaseUrl);
db.open (false);
log.info(Time taken to open connection: {}, System.currentTimeMillis()- logStartTie );
DXTickStream stream = db.getStream("bbo_stream");

CodecFactory factory = CodecFactory.newCompiledCachingFactory();
RecordClassDescriptor descriptor = stream.getFixedType();
UnboundDecoder decoder = factory.createFixedUnboundDecoder(descriptor);
MemoryDataInput in = new MemoryDataInput();
Time taken to initialize decoder: 
SelectionOptions selectionOptions = new SelectionOptions();
selectionOptions.raw = true;

try (TickCursor cursor = stream.select(startTime, selectionOptions)) {
while (cursor.next()) {
RawMessage message = (RawMessage) cursor.getMessage();

// init buffer and start reading
in.setBytes(message.data, message.offset, message.length);
decoder.beginRead(in);

// read all fields
decoder.nextField();
String trade_type = decoder.getDouble();

decoder.nextField();
double bestbidprice= decoder.getDouble();

decoder.nextField();
double bestaskprice= decoder.getDouble();

// Reading some more fields 
}
log.info(Total Time taken (opening connection and read data) to fetch records from timebase server: {}, System.currentTimeMillis()- logStartTime );
} 

I have captured timing as mentioned in the start of conversation.  Please note that at present we loaded only one day data to timebase stream which has 1 million records.

Could you please help us to improve performance?  Please let me know, if any additional information required.

**Comment by alex-karpovich** [comment]
Greetings, @bardale 

>Time taken to open connection: 8988 ms     

This is very suspicious, it is local connection or remote?

Please also remind me your java runtime version

**Reply by bardale** [reply]
Greetings, @alex-karpovich 
It's a remote connection. I am hitting TimeBase server from my laptop. Both Timebase server and laptop are in the same organizational network. 

Also could you please provide jar file as a release artifact of https://github.com/epam/TimebaseWS as its giving some compile time errors.

Thanks.

**Reply by alex-karpovich** [reply]
what is your java version?

**Reply by bardale** [reply]
openjdk version "11.0.15" 2022-04-19 LTS
openjdk Runtime Environment Zulu11.56+20-SA (build 11.0.15+10-LTS)
openjdk 64-Bit Server VM Zulu11.56+20-SA (build 11.0.15+10-LTS, mixed mode)

**Reply by alex-karpovich** [reply]
if you are familiar with TimeBase Shell CLI

https://kb.timebase.info/community/development/tools/shell

I would suggest to run next commands to check latency between server and your client
```
set db dxtick://<host>:<port>
open 
set compression auto
```

also to check throughput
```
set db dxtick://<host>:<port>
open 
set stream bbo_stream
tptime
```

**Reply by bardale** [reply]
Thanks @alex-karpovich 
My client is in a windows 10 machine.
Till yesterday I was requesting Timebase server from laptop at home via VPN.
Today I tried same request from my laptop at office. Below are stats:

Time required to fetch 30 mins timespan data:

Time taken to open connection: 3070 ms
Time taken to open stream: 120 ms
Total Time taken (opening connection and read data) to fetch records from timebase server: 4404 ms
Total number of records fetched: 17477

Time required to fetch 60 mins timespan data:

Time taken to open connection: 3024 ms
Time taken to open stream: 227 ms
Total Time taken (opening connection and read data) to fetch records from timebase server: 4802 ms
Total number of records fetched: 38318

I can see there is  improvement in timings.
Total Time taken reduced to 4404 ms from 13455 for 30 mins time span. Also for 60 min span it's taking now 4802 ms instead of 15792 ms. 
Also I can see below log messages on console.
[default-nioEventLoopGroup-1-3] Using compression: connect latency [157 ms] > 19 ms.

Still I can see it's taking time.  It's taking around 3 sec to open connection. And it's taking around 1.3 to 1.88 seconds to get messages.

Thanks.

**Reply by alex-karpovich** [reply]
Please try to check latencies using timebase shell CLI 
```
set db dxtick://<host>:<port>
open 
set compression auto
```

and share results

**Reply by bardale** [reply]
I am using windows machine so can't run tickdb.sh.

**Reply by alex-karpovich** [reply]
use tickdb.cmd

**Reply by alex-karpovich** [reply]
@bardale do you have any updates?

**Reply by bardale** [reply]
Greetings, [@alex-karpovich](https://github.com/alex-karpovich) 
Due to restrictions, I am not able to install TimeBase shell, hence unable to capture stats of tptime.
I deployed code changes to read TimeBase data on one of our environment and now getting response in less than a second including opening connection and fetching data, which much better than stats provided earlier. 
Thanks for your help.
